### PR TITLE
Add nuget package support in windows CI pipeline

### PR DESCRIPTION
**Pull request by toothache**
Add nuget package support in windows CI pipeline:
- Update CI pipeline:
  - Add build versioning;
  - Add support to make nuget packages.

### CODE REVIEW

File: win/pkgs/ort.ps1
Comment (xkszltl) [inline win/pkgs/ort.ps1]: Note that all cherry-picked commit here must be PR to upstream for review, with expectation that it'll be merged soon. And once merged, switch to use commit hash from master.

Please also add comment on PR URL.

File: win/nuget/c-ares/Roaster.c-ares.v142.dyn.x64.nuspec
Comment (xkszltl) [inline win/nuget/c-ares/Roaster.c-ares.v142.dyn.x64.nuspec]: Please move all 141->142 change into a dedicated PR.
These will be breaking change, I'll also advance Roaster version tag for it once merged.

File: win/nuget/boost/Roaster.boost.v142.dyn.x64.nuspec
Comment (xkszltl) [inline win/nuget/boost/Roaster.boost.v142.dyn.x64.nuspec]: Get some decent sentences from upstream website for description.

File: win/nuget/boost/Roaster.boost.v142.dyn.x64.targets
Comment (xkszltl) [inline win/nuget/boost/Roaster.boost.v142.dyn.x64.targets]: image
Recommend to add \n at the end.
Per unix standard, text file should end with \n.
Usually Linux editor (e.g. vim) does this automatically for you upon opening, which can cause modification in unrelated PR in the future.
On the other side, Windows editor usually won't do this and you'll have to do it manually.

File: gitlab-ci/pack.ps1
Comment (xkszltl) [inline gitlab-ci/pack.ps1]: \n

File: gitlab-ci/pack.ps1
Comment (xkszltl) [inline gitlab-ci/pack.ps1:1]: Emmm, is this the same file as win/pack.ps1?

File: azure-pipelines.windows.yml
Comment (xkszltl) [inline azure-pipelines.windows.yml]: FYI be careful when you say `cu*` for cuda, because of `curl`....

Comment (xkszltl) [reply azure-pipelines.windows.yml]: (This is a general suggestion, not only for this PR)

**Review by xkszltl — CHANGES_REQUESTED** [review]

File: win/pkgs/ort.ps1
Comment (toothache) [reply win/pkgs/ort.ps1]: I haven't got time to investigating the issue deeper. Not sure if this a really issue from ort repo because ort CI pipelines doesn't report such kind of build failure.

Will properly handle this patch when I got more information here.

File: win/nuget/c-ares/Roaster.c-ares.v142.dyn.x64.nuspec
Comment (toothache) [reply win/nuget/c-ares/Roaster.c-ares.v142.dyn.x64.nuspec]: Sure. This makes sense.

File: gitlab-ci/pack.ps1
Comment (toothache) [reply gitlab-ci/pack.ps1:1]: Not exactly. This file only prepares the folder structure. It does has many dupe codes of determining install prefix. Maybe we should bring that part as some utility functions?

File: win/pkgs/ort.ps1
Comment (xkszltl) [reply win/pkgs/ort.ps1]: We need to make sure commits are validate before cherry-pick.
For sanity of roaster packages, the earliest time we can take a patch is when upstream has taken the PR, and after discussion there's no something deeply wrong, and it at least passed their CI tests.
This is also the minimum workflow required to verify a patch itself, not only roaster related.

File: gitlab-ci/pack.ps1
Comment (xkszltl) [reply gitlab-ci/pack.ps1:1]: Then why do we need this, could you clarify the use case?

BTW I don't like the idea of having function in script, as it causes non-linear control flow.
For reusing code, switches/subscripts are usually better choice.

Comment (toothache) [reply gitlab-ci/pack.ps1:1]: like the filename suggests, this will be used in ci and it prepares desired folder structures for nuget package. win/pack.ps1 is similar with win/setup.ps1, which is target to be used locally.

and i can't agree the idea of not having function in script. The readability will be too bad. At least for powershell, I barely see scripts without any functions.

**Comment by xkszltl** [comment]
> and i can't agree the idea of not having function in script. The readability will be too bad. At least for powershell, I barely see scripts without any functions.

IMO script is a polished snapshot of terminal interaction, which is naturally linear.
It's not a programming, and I rarely see anyone writing function interactively.

There're several points to highlight here, mostly about function in script:
- Large scope non-linearity (i.e. long jump rather than branch) also hurt readability a lot.
    - Local lambda (e.g. sub-shell) is fine if closure required.
- Unlike normal programming languages, shell scripts usually have strong global status. Almost all commands can easily have global side-effects. So function lost its closure feature and becomes only a callable code snippet. Meaning you cannot rely on a symbol on caller side to understand what will happen without deep understanding of the function itself.
- You cannot import a function from shell script like what people typically did in python. So function is usually unusable outside the file without special design. On the contrary, you can call another script file anywhere, including directly from terminal.
- When implement as standalone file, we required it is (at least almost) runnable by itself. This yields much higher quality of code than a function.
- We should be (almost) POSIX compatible for portability. Currently we limit non-POSIX parts to GNU utilities and a few bash-specific syntax (e.g. `<<<`). They can be easily replaced with POSIX counter-parts.
- Powershell is quite special. `.ps1` is not exactly a snapshot of powershell command, as it has slightly different parser, and there's also function-specific `.psm1`.
    - Quality of its community (e.g. stackoverflow answers) also seems to be lower than other sh/bat with more complexity implementations.
    - It doesn't have backward-compatibility needs so it can be more aggressive on syntax.
- Then why not just have function in powershell? Because we want to keep the project portable-ready, meaning we want whatever said in powershell can be directly translated to unix/linux shell without much change.
    - This is also why we can add Windows support previously based on existing Linux code, **within a short timeframe**, when requested.

File: win/pkgs/ort.ps1
Comment (xkszltl) [reply win/pkgs/ort.ps1]: How's this going?

> Not sure if this a really issue from ort repo because ort CI pipelines doesn't report such kind of build failure.

Forgot to answer this in the first reply...
Based on our experience their CI coverage isn't good, most of issues I reported are not covered so don't be surprised.

File: win/nuget/opencv-dev/Roaster.opencv-dev.v142.dyn.x64.nuspec
Comment (xkszltl) [inline win/nuget/opencv-dev/Roaster.opencv-dev.v142.dyn.x64.nuspec]: sort

File: win/pkgs/grpc.ps1
Comment (toothache) [inline win/pkgs/grpc.ps1:93]: extra line

File: azure-pipelines/build_stage.windows.yml
Comment (xkszltl) [inline azure-pipelines/build_stage.windows.yml]: Recommend to use / for path.
\ can often be treated as escape, and won't work in anywhere outside windows.

File: win/pkgs/grpc.ps1
Comment (xkszltl) [inline win/pkgs/grpc.ps1:74]: The issue has been closed.
Is the issue still there?

File: win/pkgs/grpc.ps1
Comment (xkszltl) [inline win/pkgs/grpc.ps1]: Use / for path

Comment (xkszltl) [inline win/pkgs/grpc.ps1]: You should also specify other options like -G to take the correct compiler.

File: win/pkgs/grpc.ps1
Comment (xkszltl) [inline win/pkgs/grpc.ps1]: Do we need to build to get headers.
Is there any generated headers?

**Review by xkszltl — CHANGES_REQUESTED** [review]

File: win/pkgs/grpc.ps1
Comment (toothache) [reply win/pkgs/grpc.ps1:74]: Yes. This is closed as by design and the maintainer is suggesting to manually install the headers.

File: win/pkgs/grpc.ps1
Comment (toothache) [reply win/pkgs/grpc.ps1]: https://deploy-preview-752--grpc-io.netlify.app/docs/languages/cpp/quickstart/#build-and-install-grpc-protocol-buffers-and-abseil

File: azure-pipelines/build_stage.windows.yml
Comment (toothache) [reply azure-pipelines/build_stage.windows.yml]: yes. but this is powershell task. pwsh is also cross-platform.

File: win/pkgs/ort.ps1
Comment (toothache) [reply win/pkgs/ort.ps1]: Didn't spend more time investigating this issue. For ort specific issues, I'm not going to address in this PR.

File: azure-pipelines/build_stage.windows.yml
Comment (xkszltl) [reply azure-pipelines/build_stage.windows.yml]: Will powershell treats `a\n` as `a<newline>` or `a/n` elsewhere?

File: win/pkgs/grpc.ps1
Comment (xkszltl) [reply win/pkgs/grpc.ps1:74]: That make sense

File: win/pkgs/grpc.ps1
Comment (xkszltl) [reply win/pkgs/grpc.ps1]: \ is still there, unresolving...

File: win/pkgs/grpc.ps1
Comment (xkszltl) [reply win/pkgs/grpc.ps1]: 1. Do not copy-paste without understanding.
2. That code is doing something different. It builds and uses the libs, while here you only need the header.

File: azure-pipelines/build_stage.windows.yml
Comment (toothache) [reply azure-pipelines/build_stage.windows.yml]: we're talking about the path separator here. and powershell is using ` to as escape character

File: win/pkgs/grpc.ps1
Comment (toothache) [reply win/pkgs/grpc.ps1]: I want to install abseil header here. please read the github issue link I mentioned in `grpc.ps1`

Comment (xkszltl) [reply win/pkgs/grpc.ps1]: Yes I've read that. I think you missed the point. Unless there's header generated from template, there's nothing to build for `.h` files.
You should be able to simple copy those .h/.inc directly from source code like this:
https://github.com/abseil/abseil-cpp/blob/037ade20d1132781aae3cda4d547a9e6a5f557bf/CMakeLists.txt#L193-L200

**Comment by xkszltl** [comment]
Merging the PR for now to reduce potential conflicts since it has been there for a while.
Will work on some refactoring if I have time.
@toothache Feel free to continue on your side as well.

**Comment by xkszltl** [comment]
@toothache 
PR has been merged. Thanks for your contribution!

**Pull request merged**
